game/src/game/components/Controllers/AIController.h

```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+#include <object_ptr.hpp>
+
+#include "engine/AI/AIService.h"
+#include "engine/physics/VehicleCommands.h"  // to get the command struct
+#include "engine/scene/Component.h"
+#include "engine/scene/OnUpdateEvent.h"
+#include "engine/scene/Transform.h"
+#include "game/components/VehicleComponent.h"
+
+class AIController final : public Component,
+                           public IEventSubscriber<OnUpdateEvent>
+{
+  public:
+    AIController();
+    // From Component
+    void OnInit(const ServiceProvider& service_provider) override;
+    void OnUpdate(const Timestep& delta_time) override;
+    std::string_view GetName() const override;
+
+  private:
+    jss::object_ptr<Transform> transform_;
+    jss::object_ptr<InputService> input_service_;
+    jss::object_ptr<AIService> ai_service_;
+
+    // variables for car
+
+    std::vector<glm::vec3> path_to_follow_;
+    DirectDriveVehicle* vehicle_reference_;
+
+    Command executable_command_;
+    float timestep_ = 1.f / 60.f;
+
+    // as we want the car to move from current to next command, and so on until
+    // the end.
+    glm::vec3 next_car_position_;
+    int next_path_index_ = 2;
+
+    inline double GetEuclideanDistance(glm::vec3 first, glm::vec3 second)
+    {
+        return sqrt(pow(first.x - second.x, 2) + pow(first.y - second.y, 2) +
+                    pow(first.z - second.z, 2));
+    }
+
+  public:
+    inline void SetGVehicle(DirectDriveVehicle& vehicle)
+    {
+        vehicle_reference_ = &vehicle;
+    }
```

Comment:
In general, could we avoid switching back & forth between private/public sections all the time? If we often switch between private/public/private/public/etc it starts to get confusing as to what visibility we have for the func we're looking at. Could we maybe just have it structured like
```c++
class Thing
{
  public:
    // variables first
    int some_var_;
    // then functions
    void DoThing();
    // then maybe getters/setters
    void SetSomeVar(int some_var);

    int GetSomeVar() const;

  protected:
        // variables first
        // then functions
        // then maybe getters/setters  

  private:
        // variables first
        // then functions
        // then maybe getters/setters
};
```